fs: fatfs: additional ELM FAT options

JordanYates opened the pull request:
Make additional ELM FAT library options configurable to the user.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/35848
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/64609

JordanYates committed:
fs: fatfs: additional ELM FAT options

Make additional ELM FAT library options configurable to the user.

Signed-off-by: Jordan Yates <jordan@embeint.com>

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -88,6 +88,38 @@ config FS_FATFS_NUM_DIRS
+	  Enable retrieving and setting the filesystem labels through
+	  `f_getlabel` and `f_setlabel`.
+
+config FS_FATFS_USE_EXPAND
+	bool "Pre-allocate contiguous files"
+	help
+	  Enable pre-allocating a contiguous data area to the file through
+	  `f_expand`. Unlike expansion through `f_truncate`, `f_expand` does
+	  not write to the expanded memory area.
```
I think that this should be only available when using RAW access. Also, the option depends on `CONFIG_FATFS_READ_ONLY=n`

JordanYates replied:
Added the missing dependency to this symbol and `FS_FATFS_MKFS`.
Enabling this option without the `RAW` symbol I have proposed is not incorrect, although probably not too useful.
Either way the symbol does not exist yet so I cannot add any dependency.

JordanYates committed:
fs: fatfs: add missing dependency

Add a missing dependency on `!FS_FATFS_READ_ONLY`` for `FS_FATFS_MKFS`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -88,6 +89,39 @@ config FS_FATFS_NUM_DIRS
+	    10: 5 = Minute (0..59)
+	     4: 0 = Seconds/2 (0..29)
+
+config FS_FATFS_USE_LABEL
+	bool "Set/Get filesystem labels"
+	help
+	  Enable retrieving and setting the filesystem labels through
+	  `f_getlabel` and `f_setlabel`.
+
+config FS_FATFS_USE_EXPAND
+	bool "Pre-allocate contiguous files"
+	depends on !FS_FATFS_READ_ONLY
+	help
+	  Enable pre-allocating a contiguous data area to the file through
+	  `f_expand`. Unlike expansion through `f_truncate`, `f_expand` does
+	  not write to the expanded memory area.
+
+config FS_FATFS_USE_FIND
+	bool "Native directory iteration"
+	help
+	  Enable the native dictionary iteration functions `f_findfirst` and
+	  `f_findnext`.
+
```
Can we have these under "Direct ELM FAT interaction API" or something like that, to separate this from other things that are accessed via the VFS API?

JordanYates replied:
The added options already live under a dedicated menu:
https://github.com/zephyrproject-rtos/zephyr/blob/93c9da66944c7c7f79d5c554ca9a640b422d39b3/subsys/fs/Kconfig.fatfs#L15-L16

de-nordic replied:
Sorry, I have failed to notice
![image](https://github.com/user-attachments/assets/0451db29-bb1f-4de2-b27f-217caf7d1841)

The reason I was asking for sub-menu is that the options do not regulate how Zephyr interacts with ELM via VFS, or rather user do so, but are rather options enabling direct interaction; actually the F_USE_LABEL* has a small impact on `dir_read`, which is not clearly described in ELM driver.
Since the functions do not impact* the other parts of ELM and additional functions will only be linked when user actually calls them, then we can just have single Kconfig enabling them all. No need to multiply these.

JordanYates replied:
You touched on one option that has side effects, but `FF_HAS_RTC` requires the user to provide a function implementation, how would this be handled under a single Kconfig?

de-nordic replied:
Yes, two options: one adds the RTC, the other all the extra stuff that is not affecting the VFS usage of ELM driver.

de-nordic commented on subsys/fs/Kconfig.fatfs:
```diff
@@ -88,6 +89,39 @@ config FS_FATFS_NUM_DIRS
+	help
+	  Enable file system timestamping instead of using a hardcoded date
+	  for all operations. Requires an application supplied implementation
+	  of `get_fattime`. Format of the uint32_t bits are as follows:
```
There should be at least a test or sample providing example of reference implementation in context of Zephyr for the enabled function.

JordanYates committed:
tests: fs: fat_fs_api: add timestamp integration

Add an example implementation of `get_fattime` to the API tests. Also
ensure that `FS_FATFS_EXTRA_NATIVE_API` compiles in a test.

Signed-off-by: Jordan Yates <jordan@embeint.com>

de-nordic reviewed (approved):
Nice, thanks!
